tiled/catalog/orm.py

```diff
@@ -97,20 +100,156 @@ class Node(Timestamped, Base):
     )
 
 
-data_source_asset_association_table = Table(
-    "data_source_asset_association",
-    Base.metadata,
-    Column(
-        "data_source_id",
-        Integer,
+class DataSourceAssetAssociation(Base):
+    """
+    This describes which Assets are used by which DataSources, and how.
+
+    The 'parameter' describes which argument to pass the asset into when
+    constructing the Adapter. If 'parameter' is NULL then the asset is an
+    indirect dependency, such as a HDF5 data file backing an HDF5 'master'
+    file.
+
+    If 'num' is NULL, the asset is passed as a scalar value, and there must be
+    only one for the given 'parameter'. If 'num' is not NULL, all the assets
+    sharing the same 'parameter' (there may be one or more) will be passed in
+    as a list, ordered in ascending order of 'num'.
+    """
+
+    __tablename__ = "data_source_asset_association"
+
+    data_source_id: Mapped[int] = mapped_column(
         ForeignKey("data_sources.id", ondelete="CASCADE"),
-    ),
-    Column(
-        "asset_id",
-        Integer,
+        primary_key=True,
+    )
+    asset_id: Mapped[int] = mapped_column(
         ForeignKey("assets.id", ondelete="CASCADE"),
-    ),
-)
+        primary_key=True,
+    )
+    parameter = Column(Unicode(255), nullable=True)
+    num = Column(Integer, nullable=True)
+
+    data_source: Mapped["DataSource"] = relationship(
+        back_populates="asset_associations"
+    )
+    asset: Mapped["Asset"] = relationship(
+        back_populates="data_source_associations", lazy="selectin"
+    )
+
+    __table_args__ = (
+        UniqueConstraint(
+            "data_source_id",
+            "parameter",
+            "num",
+            name="parameter_num_unique_constraint",
+        ),
+        # Below, in unique_parameter_num_null_check, additional constraints
+        # are applied, via triggers.
+    )
+
+
+@event.listens_for(DataSourceAssetAssociation.__table__, "after_create")
+def unique_parameter_num_null_check(target, connection, **kw):
+    # Ensure that we cannot mix NULL and INTEGER values of num for
+    # a given data_source_id and parameter, and that there cannot be multiple
+    # instances of NULL.
+    if connection.engine.dialect.name == "sqlite":
+        connection.execute(
+            text(
+                """
+CREATE TRIGGER cannot_insert_num_null_if_num_int_exists
```

Comment:
I think the trigger name is misleading because it does not matter whether the existing entry is `null` or `int`...just that it already exists, right?

Perhaps something like `cannot_insert_num_null_if_num_exists`?

Reply:
Good suggestion! Updated.